### PR TITLE
refactor(pageserver): designated api error type for cancelled request

### LINKED ISSUE #7406: cancellation of get_lsn_by_timestamp handler prints full anyhow backtrace

### PR DESCRIPTION
**Pull request by yliang412**
## Problem

When a `get_lsn_by_timestamp` request is cancelled, an anyhow error is exposed to handle that case, which verbosely logs the error. However, we don't benefit from having the full backtrace provided by anyhow in this case.

## Summary of changes

This PR introduces a new `ApiError` type to handle errors caused by cancelled request more robustly.
-  A new enum variant `ApiError::Cancelled`
- Currently the cancelled request is mapped to status code 500.
- Need to handle this error in proxy's `http_util` as well.
- Added a failpoint test to simulate cancelled `get_lsn_by_timestamp` request.
 
closes #7406.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: test_runner/regress/test_lsn_mapping.py
Comment (skyzh) [inline test_runner/regress/test_lsn_mapping.py:114]: Better to add some comments that link to this pull request, or have a separate test case testing the behavior of cancellation. In general, I feel it's weird to mix cancellation test into a test case called `test_lsn_mapping`.

**Review by skyzh — COMMENTED** [review]
The test case seems failing with

```
DID NOT RAISE <class 'fixtures.pageserver.http.PageserverApiException'>
```

https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7949/9371219386/index.html#suites/2568f9ab62eb9e71321fab6263eed23e/c7a5d27333028f40

File: test_runner/regress/test_lsn_mapping.py
Comment (yliang412) [reply test_runner/regress/test_lsn_mapping.py:114]: Thanks for the catch. I forgot to actually raise the exception on Python side. I will also create a separate test for this as well.

File: test_runner/regress/test_lsn_mapping.py
Comment (skyzh) [inline test_runner/regress/test_lsn_mapping.py:138]: should be able to directly get an initial timeline ID from the env builder, but it's fine to copy other tests for now and fit it later.

File: test_runner/regress/test_lsn_mapping.py
Comment (skyzh) [inline test_runner/regress/test_lsn_mapping.py:121]: I don't think all these configurations are necessary

Comment (skyzh) [reply test_runner/regress/test_lsn_mapping.py:121]: You can take a look at `test_aux_v2_config_switch` on how to retrieve default tenant+timeline ID

**Review by skyzh — COMMENTED** [review]
LGTM and will approve once the CI passes

**Comment by skyzh** [comment]
possible reason for CI failure: CI uses vectored get path by default, so it does not hit get_value_reconstruct_data. Might need to move the failpoint to an earlier point.

File: test_runner/regress/test_lsn_mapping.py
Comment (yliang412) [reply test_runner/regress/test_lsn_mapping.py:137]: test case failed because the failpoint is not hit on the vector path. Moved the failpoint to a common path that also return `ReconstructError::Cancelled`.

File: test_runner/regress/test_lsn_mapping.py
Comment (yliang412) [reply test_runner/regress/test_lsn_mapping.py]: addressed in latest commit

File: test_runner/regress/test_lsn_mapping.py
Comment (skyzh) [inline test_runner/regress/test_lsn_mapping.py:132]: unused variable?

**Review by skyzh — APPROVED** [review]
LGTM

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:119]:
```suggestion
def test_get_lsn_by_timestamp_cancelled(neon_env_builder: NeonEnvBuilder):
    """
    Test if cancelled pageserver get_lsn_by_timestamp request is correctly handled.
    Added as an effort to improve error handling and avoid full anyhow backtrace
    """
    env = neon_env_builder.init_start()
```

IMHO no need for:
- surrounding `#`
- better as a python whatever this is called, because it will be picked up for the allure test report
- linking the PR, it will be apparent when using squash and merge in the subject of the commit

File: pageserver/src/pgdatadir_mapping.rs
Comment (koivunej) [inline pageserver/src/pgdatadir_mapping.rs:432]: I think we could move this to the top of the fn body..?

IMHO the comment could be removed, there is no other reason for such failpoint to exist.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:120]: creation of endpoints without a context manager presents an interesting test flakyness point: what if the endpoint is requesting pageserver for something which fails during shutdown.

in general, I'd prefer to see:

```suggestion
    with endpoints.create_start("main") as endpoint:
        # one more identation level
```

because then it describes the scope in which the endpoint exists and can contact safekeepers and pageserver.

however this test does not require anything to exist in the database, we only require that initial_tenant and initial_timeline are present, and we can query the database for timestamp=0 for example, it doesn't matter.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:123]: I would avoid this in new tests; I've sprinkled them around the test suite as previously there was a common method of always branching a timeline and testing that, in an effort to have a single pageserver serve multiple test cases. the python support for that was never realized with multiple processes, and it would be incompatible with global failpoints as well.

so, I'd just use these directly in a new test.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:144]: Thinking this again, I don't think this will actually reproduce the original case in the issue.

Original case in the issue was a logging "Cancelled request failed: {e:?}" or something like that.

This will only hit the failpoint, but it will not exercise the same path. In the original case, the requester was not around to wait for the response which is inspected here.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:135]: I think this is bad that we use http_client as a context manager; it does not implement those methods. I know this is copypasted from some other test case, but lets just make this a normal variable.

HTTP client by nature is stateless, I don't know what started this idea of using it as a context manager.

**Review by koivunej — CHANGES_REQUESTED** [review]
I don't think this reproduces the case in the issue.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [reply test_runner/regress/test_lsn_mapping.py:144]: One could argue we shouldn't be logging it as an `error!` when this new Cancelled variant is observed post cancellation.

**Review by khanova — APPROVED** [review]
Looks good on the proxy side.

File: libs/utils/src/http/error.rs
Comment (jorgehermo9) [inline libs/utils/src/http/error.rs:81]:
```suggestion
                self.to_string(),
```
That error message is already in the enum variant.

Sorry if I'm meddling in this PR. Just got interested by the project and ended up here

File: proxy/src/serverless/http_util.rs
Comment (jorgehermo9) [inline proxy/src/serverless/http_util.rs:49]:
```suggestion
            this.to_string(),
```
The same as my former comment

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:132]: I still think we could just remove the all of `endpoint` usage in this test, it has no meaning (we never end up actually looking at the data in the database. It would suffice to leave out the whole section considering `endpoint`.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:139]: The `conn` and `cur` are also context managers. We have a default mode of autocommit=True but I recall being confused by the different ways in different kind of uses.

However, there's a bigger problem, assuming we want to continue keeping this `endpoint` configuration:

When these statements complete, from the point of view of the system under test (compute, safekeeper, pageserver) it is indeterminate how much of the WAL (your actions with the `endpoint`) has arrived at pageserver.

Usually we use `wait_for_last_(flush|record)_lsn` to make sure the wal arrives to pageserver.

But as said, in this case we do not need any changes in the timeline database, so we could just query with timestamp 0.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:169]:
```
nextmsg, offset = wait_until(20, 0.5, lambda: end.pageserver.assert_log_contains(".*", offset))
```

the above should give us the next message, and if it would start with non-whitespace character (timestamp) then it would not be a debug formatted anyhow error. However, we don't really need to go there, because the variant now has no anyhow in it all, so I think we could remove this line as well.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:124]: This newline at the end of this line is curious.

The way python re matching works is that it matches a prefix of the string, which is a weird "match" (usually matching means that the whole string needs to match), but I am surprised that we'd need this `\n` as all lines end with that. Perhaps I would had expected `$` instead of `\n` (or neither).

**Review by arpad-m — APPROVED** [review]
looks good for me, if @koivunej 's review concerns are addressed

File: proxy/src/serverless/http_util.rs
Comment (koivunej) [reply proxy/src/serverless/http_util.rs:49]: s/this/self/ like in the other, but thanks for chiming in.

File: test_runner/regress/test_lsn_mapping.py
Comment (koivunej) [inline test_runner/regress/test_lsn_mapping.py:155]:
```suggestion
                "Cancelled request finished with an error: Cancelled$", offset
```

`$` will only match at the end of the line (`\n`) so there is nothing coming after that, we could just leave it off.

**Review by koivunej — APPROVED** [review]

**Pull request merged**
